File: Python/OneLang/Test/TestCases/ProjectGeneratorTest.py
from onelang_core import *
from onelang_file import *
import OneLang.Test.TestCase as testCase
import OneLang.One.CompilerHelper as compHelp
import OneLang.One.Compiler as comp
import OneLang.Test.PackageStateCapture as packStateCapt
import OneLang.Generator.ProjectGenerator as projGen

class StageExporter:
    def __init__(self, artifact_dir, compiler):
        self.stage = 0
        self.artifact_dir = artifact_dir
        self.compiler = compiler
    
    def after_stage(self, stage_name):
        console.log(f'''Stage finished: {stage_name}''')
        OneFile.write_text(f'''{self.artifact_dir}/stages/{self.stage}_{stage_name}.txt''', packStateCapt.PackageStateCapture(self.compiler.project_pkg).get_summary())
        self.stage = self.stage + 1

class ProjectGeneratorTest:
    def __init__(self, base_dir):
        self.name = "ProjectGeneratorTest"
        self.base_dir = base_dir
    
    def get_test_cases(self):
        return [testCase.TestCase("OneLang", lambda artifact_dir: self.compile_one_lang(artifact_dir))]
    
    def compile_one_lang(self, artifact_dir):
        console.log("Initalizing project generator...")
        proj_gen = projGen.ProjectGenerator(self.base_dir, f'''{self.base_dir}/xcompiled-src''')
        proj_gen.out_dir = f'''{artifact_dir}/output/'''
        
        console.log("Initalizing project for compiler...")
        compiler = compHelp.CompilerHelper.init_project(proj_gen.project_file.name, proj_gen.src_dir, proj_gen.project_file.source_lang, None)
        compiler.hooks = StageExporter(artifact_dir, compiler)
        
        console.log("Processing workspace...")
        compiler.process_workspace()
        
        console.log("Generating project...")
        proj_gen.generate()

File: PHP/src/OneLang/Test/TestCases/ProjectGeneratorTest.php
<?php

namespace OneLang\Test\TestCases\ProjectGeneratorTest;

use OneLang\File\OneFile;
use OneLang\Test\TestCase\ITestCollection;
use OneLang\Test\TestCase\TestCase;
use OneLang\One\CompilerHelper\CompilerHelper;
use OneLang\One\Compiler\Compiler;
use OneLang\One\Compiler\ICompilerHooks;
use OneLang\Test\PackageStateCapture\PackageStateCapture;
use OneLang\Generator\ProjectGenerator\ProjectGenerator;

class StageExporter implements ICompilerHooks {
    public $stage = 0;
    public $artifactDir;
    public $compiler;
    
    function __construct($artifactDir, $compiler) {
        $this->artifactDir = $artifactDir;
        $this->compiler = $compiler;
    }
    
    function afterStage($stageName) {
        \OneLang\Core\console::log("Stage finished: " . $stageName);
        OneFile::writeText($this->artifactDir . "/stages/" . $this->stage . "_" . $stageName . ".txt", (new PackageStateCapture($this->compiler->projectPkg))->getSummary());
        $this->stage++;
    }
}

class ProjectGeneratorTest implements ITestCollection {
    public $name = "ProjectGeneratorTest";
    public $baseDir;
    
    function __construct($baseDir) {
        $this->baseDir = $baseDir;
    }
    
    function getTestCases() {
        return array(new TestCase("OneLang", function ($artifactDir) { return $this->compileOneLang($artifactDir); }));
    }
    
    function compileOneLang($artifactDir) {
        \OneLang\Core\console::log("Initalizing project generator...");
        $projGen = new ProjectGenerator($this->baseDir, $this->baseDir . "/xcompiled-src");
        $projGen->outDir = $artifactDir . "/output/";
        
        \OneLang\Core\console::log("Initalizing project for compiler...");
        $compiler = CompilerHelper::initProject($projGen->projectFile->name, $projGen->srcDir, $projGen->projectFile->sourceLang, null);
        $compiler->hooks = new StageExporter($artifactDir, $compiler);
        
        \OneLang\Core\console::log("Processing workspace...");
        $compiler->processWorkspace();
        
        \OneLang\Core\console::log("Generating project...");
        $projGen->generate();
    }
}
